File: frontend/src/api/classroom.ts
import { apiClient } from './client';

export interface ClassCreatePayload {
  name: string;
  college?: string;
  year?: string;
  semester?: string;
  regulation?: string;
  section?: string;
}

export interface ClassJoinPayload {
  code: string;
}

export interface ClassMember {
  id: number;
  student_id: number;
  student_name: string;
  student_email: string;
  joined_at: string;
}

export interface Classroom {
  id: number;
  teacher_id: number;
  teacher_name?: string;
  name: string;
  code: string;
  college?: string;
  year?: string;
  semester?: string;
  regulation?: string;
  section?: string;
  is_active: boolean;
  member_count: number;
  created_at: string;
}

export const classroomApi = {
  createClass: async (payload: ClassCreatePayload): Promise<Classroom> => {
    const response = await apiClient.post<Classroom>('/classes', payload);
    return response.data;
  },

  getTeacherClasses: async (): Promise<Classroom[]> => {
    const response = await apiClient.get<Classroom[]>('/classes/teacher');
    return response.data;
  },

  joinClass: async (payload: ClassJoinPayload): Promise<Classroom> => {
    const response = await apiClient.post<Classroom>('/classes/join', payload);
    return response.data;
  },

  getStudentClasses: async (): Promise<Classroom[]> => {
    const response = await apiClient.get<Classroom[]>('/classes/student');
    return response.data;
  },

  getClassDetails: async (classId: number): Promise<Classroom> => {
    const response = await apiClient.get<Classroom>(`/classes/${classId}`);
    return response.data;
  },

  getClassMembers: async (classId: number): Promise<ClassMember[]> => {
    const response = await apiClient.get<ClassMember[]>(`/classes/${classId}/members`);
    return response.data;
  },

  leaveClass: async (classId: number): Promise<void> => {
    await apiClient.delete(`/classes/${classId}/leave`);
  },
};

File: backend/app/schemas/classroom.py
from datetime import datetime
from typing import List, Optional
from pydantic import BaseModel, ConfigDict, Field


class ClassCreateRequest(BaseModel):
    name: str = Field(..., min_length=1, max_length=255, description="Class Name")
    college: Optional[str] = None
    year: Optional[str | int] = None
    semester: Optional[str | int] = None
    regulation: Optional[str | int] = None
    section: Optional[str] = None


class ClassJoinRequest(BaseModel):
    code: str = Field(..., min_length=1, max_length=20, description="Class Code")


class ClassMemberStudentResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    student_id: int
    student_name: str
    student_email: str
    joined_at: datetime


class ClassroomResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    teacher_id: int
    teacher_name: Optional[str] = None
    name: str
    code: str
    college: Optional[str] = None
    year: Optional[str] = None
    semester: Optional[str] = None
    regulation: Optional[str] = None
    section: Optional[str] = None
    is_active: bool
    member_count: int = 0
    created_at: datetime
